pages/3_churn_analysis.py

```diff
@@ -18,45 +21,284 @@
+                       (data['subscription_period_started_at'] <= end_date)]
+
+# Debug: Print the shape of subscribed_data
+st.write(f"Number of subscribed records: {subscribed_data.shape[0]}")
+
+# Prepare cohort data
+cohort = subscribed_data.groupby('subscription_id').agg({
+    'customer_created_at': 'first',
+    'subscription_period_started_at': 'first',
+    'subscription_status': 'last'
+}).reset_index()
+
+# Calculate months since customer creation
+cohort['months_since_customer_creation'] = ((cohort['subscription_period_started_at'].dt.to_period('M') - 
+                                             cohort['customer_created_at'].dt.to_period('M'))).apply(lambda x: max(0, x.n))
+
+# Count churns
+churn_matrix = cohort[cohort['subscription_status'] == 'inactive'].groupby([
+    cohort['subscription_period_started_at'].dt.to_period('M'), 
+    'months_since_customer_creation'
+]).size().unstack(fill_value=0)
+
+# Sort the columns and index
+churn_matrix = churn_matrix.sort_index()
+churn_matrix = churn_matrix.reindex(sorted(churn_matrix.columns), axis=1)
+
+# Rename the index for clarity
+churn_matrix.index = churn_matrix.index.strftime('%Y-%m')
+churn_matrix.index.name = 'Subscription Start Month'
+
+# Reset the index to make 'Subscription Start Month' a column
+churn_matrix_reset = churn_matrix.reset_index()
+
+# Set 'Subscription Start Month' as the first column name
+churn_matrix_reset.columns.name = None
+churn_matrix_reset = churn_matrix_reset.rename(columns={'': 'Subscription Start Month'})
+
+# Function to apply color gradient
+def color_gradient(val):
+    if isinstance(val, str):  # Don't apply color to 'Subscription Start Month' column
+        return ''
+    color = "#306BEA"
+    return f'background-color: rgba{tuple(int(color[i:i+2], 16) for i in (1, 3, 5)) + (val/churn_matrix.max().max(),)}'
+
+# Apply styling
+styled_churn_matrix = churn_matrix_reset.style.applymap(color_gradient)
+
+# Convert the styled dataframe to HTML
+html_table = styled_churn_matrix.to_html(index=False)
+
+# Add the "Months Since Customer Creation" header, adjust table CSS, and add horizontal scroll
+html_table = f"""
+<style>
+    .table-container {{
+        width: 100%;
+        overflow-x: auto;
+    }}
+    table {{
+        border-collapse: collapse;
+        width: max-content;
+    }}
+    th, td {{
+        text-align: center;
+        padding: 8px;
+        border: 1px solid #ddd;
+        min-width: 60px;
+    }}
+    th:first-child, td:first-child {{
+        text-align: left;
+        position: sticky;
+        left: 0;
+        background-color: white;
+        z-index: 1;
+    }}
+</style>
+<div style="text-align: center; font-weight: bold; margin-bottom: 10px;">
+    Months Since Customer Creation
+</div>
+<div class="table-container">
+    {html_table}
+</div>
+"""
+
+# Display the churn matrix as a table
+st.write("Churn Matrix:")
+st.markdown(html_table, unsafe_allow_html=True)
```

Comment:
What does the "unsafe_allow_html=true" mean here? Is there a way to achieve the desired matrix without using "unsafe" html?

Reply:
Additionally the Churn matrix looks very cool, but it is quite large. Is there a way we can scale it down to be more digestible? Additionally, can we include the color component to the matrix like we have in the Figma example?

Reply:
I swapped out the unsafe_allow_html, apparently Streamlit has a built in component!

Reply:
I tried a few things to get it to scale down but I feel like this matrix is probably the most fickle. I couldn't get the sizing to really change without something else changing like the scrolling or the displays so I feel stuck there. 

As for the color, I applied the gradient but if it's zero, it's white, is that what you were talking about?

Also, I'm not sure if it's just because of the dummy data, but I feel the churn rates themselves look a little off, so that may need to be revisited as well

Reply:
I also realized we should be doing rates, so I added that in (and include the absolute counts on the side). let me know what you think. i tried making the counts visually smaller than the rates, but couldn't figure that out either.